bundles/org.openhab.binding.metofficedatahub/src/main/java/org/openhab/binding/metofficedatahub/internal/MetOfficeDataHubBridgeHandler.java

```diff
@@ -0,0 +1,125 @@
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.metofficedatahub.internal;
+
+import static org.openhab.binding.metofficedatahub.internal.MetOfficeDataHubBindingConstants.BRIDGE_PROP_FORECAST_REQUEST_COUNT;
+import static org.openhab.binding.metofficedatahub.internal.MetOfficeDataHubBindingConstants.DAY_IN_MILLIS;
+
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.binding.BaseBridgeHandler;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MetOfficeDataHubBridgeHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author David Goodyear - Initial contribution
+ */
+@NonNullByDefault
+public class MetOfficeDataHubBridgeHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(MetOfficeDataHubBridgeHandler.class);
+    private final Object timerResetSchedulerLock = new Object();
+    protected final RequestLimiter forecastDataLimiter = new RequestLimiter();
+
+    private volatile MetOfficeDataHubBridgeConfiguration config = getConfigAs(
+            MetOfficeDataHubBridgeConfiguration.class);
+
+    private @Nullable ScheduledFuture<?> timerResetScheduler = null;
+
+    protected String getApiKey() {
+        return config.siteApiKey;
+    }
+
+    public MetOfficeDataHubBridgeHandler(final Bridge bridge) {
+        super(bridge);
+    }
+
+    public void updateLimiterStats() {
+        final Map<String, String> newProps = new HashMap<>();
+        newProps.put(BRIDGE_PROP_FORECAST_REQUEST_COUNT, String.valueOf(forecastDataLimiter.getCurrentRequestCount()));
+        this.updateProperties(newProps);
+    }
+
+    private static long getMillisUntilMidnight() {
+        return Duration.between(LocalDateTime.now(), LocalDate.now().plusDays(1).atStartOfDay()).toMillis();
+    }
+
+    protected static long getMillisSinceDayStart() {
+        return Duration.between(LocalDate.now().atStartOfDay(), LocalDateTime.now()).toMillis();
+    }
+
+    @Override
+    public void initialize() {
+        updateLimiterStats();
+        config = getConfigAs(MetOfficeDataHubBridgeConfiguration.class);
+        forecastDataLimiter.updateLimit(config.siteRateDailyLimit);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduler.execute(() -> updateStatus(ThingStatus.ONLINE));
```

Comment:
Yeah I did react to this, and then figured it was done like this as the relevant thing under the bridge would check the credentials when the activate. However, as it would only be a single call during initalisation to check the credential's, i am now thinking actually, its just a single call so it why not just use one extra call to validate the credentials when init is called. Will come back to this last after the other bits.

Reply:
Hi @lsiepel, this has been implemented now. I had a bit of fun with scheduler.execute() appearing to block execution, so you'll see there's a 1 second delayed execution now applied to both initialise methods to ensure this behaviour isn't seen, as it then doesn't I presume try to use the calling thread. I would like to test the code base a bit more over the next few days, so will re-request a review, but if everything is closed can we hold off the merge until next weekend, so I have some time to re-validate everything on testbeds here? (I've probably almost hit my API call limits today, so will need to stop now irrelevant, but hopefully all the changes look good 🤞). Thanks again for all the time and help on the PR's it feel's like its all in a much better state, due to all the time you've given towards it.

Reply:
After it is ready to merge, i'll certainly ask you to confirm tests before i actually merge the PR.